src/spyglass/common/common_dandi.py

```diff
@@ -0,0 +1,253 @@
+
+try:
+    import dandi.download
+    import dandi.organize
+    import dandi.upload
+    import dandi.validate
+    from dandi.consts import known_instances
+    from dandi.dandiapi import DandiAPIClient
+    from dandi.metadata.nwb import get_metadata
+    from dandi.organize import OrganizeInvalid
+    from dandi.validate_types import Severity
+
+except ImportError as e:
+    logger.warning(e)
+
+
+from spyglass.common.common_usage import Export
+from spyglass.settings import export_dir
+from spyglass.utils import SpyglassMixin, logger
+
+schema = dj.schema("common_dandi")
+
+
+@schema
+class DandiPath(SpyglassMixin, dj.Manual):
+    definition = """
+    -> Export.File
+    ---
+    dandiset_id: varchar(16)
+    filename: varchar(255)
+    dandi_path: varchar(255)
+    dandi_instance = "dandi": varchar(32)
+    """
+
+    def fetch_file_from_dandi(self, key: dict):
+        dandiset_id, dandi_path, dandi_instance = (self & key).fetch1(
+            "dandiset_id", "dandi_path", "dandi_instance"
+        )
+        dandiset_id = str(dandiset_id)
+        # get the s3 url from Dandi
+        with DandiAPIClient(
+            dandi_instance=known_instances[dandi_instance],
+        ) as client:
+            asset = client.get_dandiset(dandiset_id).get_asset_by_path(
+                dandi_path
+            )
+            s3_url = asset.get_content_url(follow_redirects=1, strip_query=True)
+
+        # stream the file from s3
+        # first, create a virtual filesystem based on the http protocol
+        fs = fsspec.filesystem("http")
+
+        # create a cache to save downloaded data to disk (optional)
+        fsspec_file = CachingFileSystem(
+            fs=fs,
+            cache_storage=f"{export_dir}/nwb-cache",  # Local folder for the cache
+        )
+
+        # Open and return the file
+        fs_file = fsspec_file.open(s3_url, "rb")
+        io = pynwb.NWBHDF5IO(file=h5py.File(fs_file))
+        nwbfile = io.read()
+        return (io, nwbfile)
+
+    def compile_dandiset(
+        self,
+        key: dict,
+        dandiset_id: str,
+        dandi_api_key: str = None,
+        dandi_instance: str = "dandi",
+    ):
+        """Compile a Dandiset from the export.
+        Parameters
+        ----------
+        key : dict
+            ExportSelection key
+        dandiset_id : str
+            Dandiset ID generated by the user on the dadndi server
+        dandi_api_key : str, optional
+            API key for the dandi server. Optional if the environment variable
+            DANDI_API_KEY is set.
+        dandi_instance : What instance of Dandi the dandiset is on. Defaults to the dev server
+        """
+        key = (Export & key).fetch1("KEY")
+        paper_id = (Export & key).fetch1("paper_id")
+        if self & key:
+            raise ValueError(
+                f"Adding new files to an existing dandiset is not permitted. Please delete existing entries for {key} and run again"
```

Comment:
In general, I try to put the variable at the end so that the message is still readable even if the variable is tough to read because length of unexpected values

```suggestion
                "Adding new files to an existing dandiset is not permitted. "
                + f"Please rerun after deleting existing entries for {key}"
```